### PR TITLE
Add hook actionExportOrder to modify the csv generated by the order export

### LINKED ISSUE #27885: The columns of the csv export from the order list is frozen
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

When setting up a site, you can add or remove columns with the hooks actionOrderGridDefinitionModifier and ActionOrderGridQueryBuilderModifier. The SQL query includes the added columns but the csv export is frozen and does not take into account the new columns

### Describe the solution you'd like

I suggest adding a new hook into exportAction

### Alternatives you've considered

_No response_

### Steps to reproduce the behavior

1. Create new table `app_order_table` with `id_order`
image

2. Install the module [ps_qualityassurance](https://github.com/PrestaShop/ps_qualityassurance), add two hooks:

`actionOrderGridQueryBuilderModifier` with this content:
```php
$searchQueryBuilder = $params['search_query_builder'];
$countQueryBuilder = $params['count_query_builder'];

    $searchCriteria = $params['search_criteria'];

    $searchQueryBuilder->addSelect(
        'app.`id_order` IS NOT NULL AS `from_app`'
    );

    $searchQueryBuilder->leftJoin(
        'o',
        '`app_order_table`',
        'app',
        'app.`id_order` = o.`id_order`'
    );
	$countQueryBuilder->leftJoin(
        'o',
        '`app_order_table`',
        'app',
        'app.`id_order` = o.`id_order`'
    );

    if ('from_app' === $searchCriteria->getOrderBy()) {
        $searchQueryBuilder->orderBy('app.`id_order`', $searchCriteria->getOrderWay());
    }

    foreach ($searchCriteria->getFilters() as $filterName => $filterValue) {
        if ('from_app' === $filterName) {
            $searchQueryBuilder
              ->andWhere('app.`id_order` IS ' . ($filterValue ? 'NOT' : '') . ' NULL');
          $countQueryBuilder
              ->andWhere('app.`id_order` IS ' . ($filterValue ? 'NOT' : '') . ' NULL');
        }
    }
```

and `actionOrderGridDefinitionModifier` with this content:
```php
$definition = $params['definition'];

    $definition->getColumns()->addAfter(
        'date_add',
        (new PrestaShop\PrestaShop\Core\Grid\Column\Type\BooleanColumn('from_app'))
            ->setName($this->l('App order'))
            ->setOptions([
                'field' => 'from_app',
                'true_name' => $this->trans('Yes', [], 'Admin.Global'),
                'false_name' => $this->trans('No', [], 'Admin.Global'),
                'clickable' => true,
            ])
    );

    $definition->getFilters()->add(
        (new PrestaShop\PrestaShop\Core\Grid\Filter\Filter('from_app', PrestaShopBundle\Form\Admin\Type\YesAndNoChoiceType::class))
        ->setTypeOptions([
            'required' => false,
        ])    
        ->setAssociatedColumn('from_app')
    );
```

3. Now, go on the page Order and see the new column 
image
4. Click export to CSV
5. See the CSV file > the new column is not displayed 
image

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by olecorre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | Allows you to modify the csv generated by the order export in the administration
| Type?             | new feature
| Category?         | BO 
| BC breaks?        | no
| Deprecations?     |  no
| Fixed ticket?     | Fixes #27885
| How to test?      | Please follow ticket instructions in https://github.com/PrestaShop/PrestaShop/issues/27885. Module [ps_qualityassurance](https://github.com/PrestaShop/ps_qualityassurance) can be used to add a column in the list of orders in the BO, make a csv export. Today the new column is missing. With this PR which inserts a new hook, you can redefine the exported columns in the csv, the data is already the SQL query
| Possible impacts? | No impact

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27886)

### CODE REVIEW

**Comment by olecorre** [comment]
Exemple to use the new hook : 
public function hookactionExportOrder($params){
        $newheader = [];

        foreach($params['header'] as $key=>$head){
            if( $key=='new' || $key=='reference') continue;
            $newheader[$key] = $head;
            if($key=='customer'){
                $newheader['products'] = $this->trans('Products',  array(),'Modules.oh_extend.Admin');
                $newheader['store'] = $this->trans('Store',  array(),'Modules.oh_extend.Admin');
                $newheader['date_livraison'] = $this->trans('Date livraison',  array(),'Modules.oh_extend.Admin');
            }
        }
        $params['header'] = $newheader;

        $data = [];

        foreach ( $params['orderGrid']->getData()->getRecords()->all() as $record) {
            $item = [
                'id_order' => $record['id_order'],
                'country_name' => $record['country_name'],
                'customer' => $record['customer'],
                'products' => $record['products'],
                'store' => $record['store'],
                'date_livraison' => $record['date_livraison'],
                'total_paid_tax_incl' => $record['total_paid_tax_incl'],
                'payment' => $record['payment'],
                'osname' => $record['osname'],
                'date_add' => $record['date_add'],
            ];

            if ($params['isB2bEnabled']) {
                $item['company'] = $record['company'];
            }

            $data[] = $item;
        }
        $params['data'] = $data;
    }

**Comment by ghost** [comment]
Nice hook, don't forget the PR in https://github.com/PrestaShop/autoupgrade

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php [hunk 1]
```diff
@@ -397,7 +397,15 @@ public function exportAction(OrderFilters $filters)
...
-
+        $this->dispatchHook(
+            'actionExportOrder',
+            [
+                'header' => &$headers,
+                'data' => &$data,
+                'orderGrid' => $orderGrid,
+                'isB2bEnabled' => $isB2bEnabled
+            ]
+        );
```
Comment (saulaski) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php]: Please fix code style
```suggestion
                'isB2bEnabled' => $isB2bEnabled,
            ]
        );

```

**Review by saulaski — CHANGES_REQUESTED** [review]

File: src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php [hunk 2]
```diff
@@ -397,7 +397,15 @@ public function exportAction(OrderFilters $filters)
...
-
+        $this->dispatchHook(
+            'actionExportOrder',
+            [
+                'header' => &$headers,
+                'data' => &$data,
+                'orderGrid' => $orderGrid,
+                'isB2bEnabled' => $isB2bEnabled,
+            ]
+        );
```
Comment (saulaski) [inline src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:409]: Please don't forget the extra line before the return statement. This is a code style rule.

You can validate the code style yourself before you push with this command line:
`php vendor/bin/php-cs-fixer fix --dry-run --diff`

**Review by saulaski — CHANGES_REQUESTED** [review]

**Comment by saulaski** [comment]
@olecorre Please rebase your commits: 1 commit / PR

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by ghost** [comment]
I think if one PR = one commit we can [merge this PR](https://github.com/PrestaShop/traces/pull/24) by @PrestaEdit 

And use the stats since 2017

**Review by matks — DISMISSED** [review]
@olecorre Apologies for the delay I have no idea why this PR was off the radar for so long.

**Comment by matks** [comment]
@olecorre A git rebase was needed. As the review came very late I did it for you.

**Review by kpodemski — APPROVED** [review]

**Review by khouloudbelguith — COMMENTED** [review]
Hi @olecorre,

I followed these steps:
1. Create new table `app_order_table` with `id_order`
image

2. Install the module [ps_qualityassurance](https://github.com/PrestaShop/ps_qualityassurance), add two hooks:

`actionOrderGridQueryBuilderModifier` with this content:
```php
$searchQueryBuilder = $params['search_query_builder'];
$countQueryBuilder = $params['count_query_builder'];

    $searchCriteria = $params['search_criteria'];

    $searchQueryBuilder->addSelect(
        'app.`id_order` IS NOT NULL AS `from_app`'
    );

    $searchQueryBuilder->leftJoin(
        'o',
        '`app_order_table`',
        'app',
        'app.`id_order` = o.`id_order`'
    );
	$countQueryBuilder->leftJoin(
        'o',
        '`app_order_table`',
        'app',
        'app.`id_order` = o.`id_order`'
    );

    if ('from_app' === $searchCriteria->getOrderBy()) {
        $searchQueryBuilder->orderBy('app.`id_order`', $searchCriteria->getOrderWay());
    }

    foreach ($searchCriteria->getFilters() as $filterName => $filterValue) {
        if ('from_app' === $filterName) {
            $searchQueryBuilder
              ->andWhere('app.`id_order` IS ' . ($filterValue ? 'NOT' : '') . ' NULL');
          $countQueryBuilder
              ->andWhere('app.`id_order` IS ' . ($filterValue ? 'NOT' : '') . ' NULL');
        }
    }
```
3. Install this module 
[testhookactionExportOrder.zip](https://github.com/PrestaShop/PrestaShop/files/8563177/testhookactionExportOrder.zip)
which uses this function 
```php
 public function hookactionExportOrder($params)
    {
        $newheader = [];

    foreach($params['header'] as $key=>$head){
        if( $key=='new' || $key=='reference') continue;
        $newheader[$key] = $head;
        if($key=='customer'){
            $newheader['products'] = $this->trans('Products',  array(),'Modules.oh_extend.Admin');
            $newheader['store'] = $this->trans('Store',  array(),'Modules.oh_extend.Admin');
            $newheader['date_livraison'] = $this->trans('Date livraison',  array(),'Modules.oh_extend.Admin');
        }
    }
    $params['header'] = $newheader;

    $data = [];

    foreach ( $params['orderGrid']->getData()->getRecords()->all() as $record) {
        $item = [
            'id_order' => $record['id_order'],
            'country_name' => $record['country_name'],
            'customer' => $record['customer'],
            'products' => $record['products'],
            'store' => $record['store'],
            'date_livraison' => $record['date_livraison'],
            'total_paid_tax_incl' => $record['total_paid_tax_incl'],
            'payment' => $record['payment'],
            'osname' => $record['osname'],
            'date_add' => $record['date_add'],
        ];

        if ($params['isB2bEnabled']) {
            $item['company'] = $record['company'];
        }

        $data[] = $item;
    }
    $params['data'] = $data;
    }
```
4. Go to BO > Orders > Orders list page
5. Click on export
6. See exception 
image

I tried to remove those lines
```php
            'country_name' => $record['country_name'],
            'customer' => $record['customer'],
            'products' => $record['products'],
```

I don't have any exceptions but the csv file is missing
image

Can you please provide the exact steps to check that the issue is fixed?

Thanks!

**Comment by olecorre** [comment]
My function :

```
public function hookactionExportOrder($params){
        $newheader = [];

        foreach($params['header'] as $key=>$head){
            if( $key=='new' || $key=='reference') continue;
            $newheader[$key] = $head;
            if($key=='customer'){
                $newheader['products'] = $this->trans('Products',  array(),'Modules.oh_extend.Admin');
                $newheader['store'] = $this->trans('Store',  array(),'Modules.oh_extend.Admin');
                $newheader['date_livraison'] = $this->trans('Date livraison',  array(),'Modules.oh_extend.Admin');
            }
        }
        $params['header'] = $newheader;

        $data = [];

        foreach ( $params['orderGrid']->getData()->getRecords()->all() as $record) {
            $item = [
                'id_order' => $record['id_order'],
                'country_name' => $record['country_name'],
                'customer' => $record['customer'],
                'products' => $record['products'],
                'store' => $record['store'],
                'date_livraison' => $record['date_livraison'],
                'total_paid_tax_incl' => $record['total_paid_tax_incl'],
                'payment' => $record['payment'],
                'osname' => $record['osname'],
                'date_add' => $record['date_add'],
            ];

            if ($params['isB2bEnabled']) {
                $item['company'] = $record['company'];
            }

            $data[] = $item;
        }
        $params['data'] = $data;
    }
```

**Review by khouloudbelguith — COMMENTED** [review]
Hi @olecorre,

Could you please check the Conflicting files?

Thanks!

**Comment by matks** [comment]
Hi all,

Since this PR has been stale since May, I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale.

**Feel free to reopen or open another one if you think it's still relevant.** 😄 

Thanks!

**Pull request closed without merging**